roachtest: inbox received batch larger than max batch size

roachtest.sqlsmith/setup=tpcc/setting=no-ddl [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6260325?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/6260325?buildTab=artifacts#/sqlsmith/setup=tpcc/setting=no-ddl) on master @ [770ff3c545a51752490403da64d56fb397f49c5e](https://github.com/cockroachdb/cockroach/commits/770ff3c545a51752490403da64d56fb397f49c5e):


```
test artifacts and logs in: /artifacts/sqlsmith/setup=tpcc/setting=no-ddl/run_1
	sqlsmith.go:267,sqlsmith.go:327,test_runner.go:897: error: pq: internal error: unexpected error from the vectorized engine: runtime error: index out of range [2012] with length 2012
		stmt:
		SELECT
			tab_105.o_carrier_id AS col_223,
			(-6454220902618163273):::INT8 AS col_224,
			tab_105.crdb_internal_mvcc_timestamp AS col_225,
			tab_105.o_carrier_id AS col_226,
			tab_105.o_entry_d AS col_227,
			tab_105.o_w_id AS col_228
		FROM
			defaultdb.public."order"@order_idx AS tab_105
		ORDER BY
			tab_105.crdb_internal_mvcc_timestamp;
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=tpcc/setting=no-ddl.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-19068

--- a/pkg/cmd/roachtest/tests/sqlsmith.go
+++ b/pkg/cmd/roachtest/tests/sqlsmith.go
@@ -21,12 +21,10 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
-	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/registry"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/spec"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
 	"github.com/cockroachdb/cockroach/pkg/internal/sqlsmith"
-	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/errors"
 )
@@ -94,13 +92,12 @@ func registerSQLSmith(r registry.Registry) {
 		rng, seed := randutil.NewTestRand()
 		t.L().Printf("seed: %d", seed)
 
-		// With 50% chance use the cockroach-short binary that was compiled with
-		// --crdb_test build tag.
-		maybeUseBuildWithEnabledAssertions(ctx, t, c, rng, 0.5)
 		if err := c.PutLibraries(ctx, "./lib"); err != nil {
 			t.Fatalf("could not initialize libraries: %v", err)
 		}
-		c.Start(ctx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings())
+		// With 50% chance use the cockroach-short binary that was compiled with
+		// --crdb_test build tag.
+		maybeUseBuildWithEnabledAssertions(ctx, t, c, rng, 0.5 /* eaProb */)
 
 		setupFunc, ok := setups[setupName]
 		if !ok {

--- a/pkg/cmd/roachtest/tests/util.go
+++ b/pkg/cmd/roachtest/tests/util.go
@@ -18,7 +18,9 @@ import (
 	"time"
 
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
+	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/option"
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/test"
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
 	"github.com/cockroachdb/cockroach/pkg/util/retry"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/errors"
@@ -129,17 +131,29 @@ func SetAdmissionControl(ctx context.Context, t test.Test, c cluster.Cluster, en
 
 // maybeUseBuildWithEnabledAssertions stages the cockroach-short binary with
 // enabled assertions with eaProb probability if that binary is available,
-// otherwise stages the regular cockroach binary.
+// otherwise stages the regular cockroach binary, and starts the cluster.
 func maybeUseBuildWithEnabledAssertions(
 	ctx context.Context, t test.Test, c cluster.Cluster, rng *rand.Rand, eaProb float64,
 ) {
 	if rng.Float64() < eaProb {
 		// Check whether the cockroach-short binary is available.
 		if t.CockroachShort() != "" {
-			t.Status("using cockroach-short binary compiled with --crdb_test build tag")
+			randomSeed := rng.Int63()
+			t.Status(
+				"using cockroach-short binary compiled with --crdb_test "+
+					"build tag and COCKROACH_RANDOM_SEED=", randomSeed,
+			)
 			c.Put(ctx, t.CockroachShort(), "./cockroach")
+			// We need to ensure that all nodes in the cluster start with the
+			// same random seed (if not, some assumptions can be violated - for
+			// example that coldata.BatchSize() values are the same on all
+			// nodes).
+			settings := install.MakeClusterSettings()
+			settings.Env = append(settings.Env, fmt.Sprintf("COCKROACH_RANDOM_SEED=%d", randomSeed))
+			c.Start(ctx, t.L(), option.DefaultStartOpts(), settings)
 			return
 		}
 	}
 	c.Put(ctx, t.Cockroach(), "./cockroach")
+	c.Start(ctx, t.L(), option.DefaultStartOpts(), install.MakeClusterSettings())
 }

